data/common.yaml

```diff
@@ -1 +1,134 @@
---- {}
+---
+profile_audit::qualys::enabled: true
+profile_audit::qualys::gid: "19999"
+profile_audit::qualys::group: "qualys"
+profile_audit::qualys::homedir: "/home/qualys"
+profile_audit::qualys::ip: "141.142.148.51"
+profile_audit::qualys::ssh_authorized_key: null
+profile_audit::qualys::ssh_authorized_key_type: "rsa"
+profile_audit::qualys::uid: "19999"
+profile_audit::qualys::user: "qualys"
+profile_audit::qualys::user_comment: "NCSA IRST Qualys - security@ncsa.illinois.edu"
+
+profile_audit::root_equivalence::crons:
+  "root equivalence reporting":
+    command: "/root/scripts/k5login_otp_audit.pl -clear > /dev/null 2>&1"
+    environment:
+      - "SHELL=/bin/sh"
+    hour: 1
+    minute: 1
+    monthday: 1
+    user: "root"
+profile_audit::root_equivalence::files:
+  "/root/scripts":
+    ensure: "directory"
+    mode: "0750"
+    owner: "root"
+    group: "root"
+  "/root/scripts/k5login_otp_audit.pl":
```

Comment:
Can this be moved to the module "files" directory?

Reply:
I've purposely been trying to stop putting scripts like this under `./files/` because then there is no easy way to overwrite them outside of the module. Convince me why we would want to hardcode this under `./files/`. ;)

Reply:
To me, the point of making profiles is to keep the high level view of machine configuration simple by encapsulating the details. Putting the entire contents of a script into hiera doesn't fit in line with this. Rather, it merely moves complexity into hiera (instead of encapsulating it.)

I see your point about "no easy way to change the file", but that begs the question, what needs to change?

Perhaps the file could be updated so that it takes parameters for any parts that are changing per role/location/project. Then just those parameters can go into hiera.

Alternatively, if the script has significant changes based on role/location/project (or other hiera level distinction), perhaps there should be multiple scripts and the one that get's put in place is determined by a class parameter. I like this idea less because making a script change necessitates reviewing each script, which provides more opportunity for error and also requires more testing.

Reply:
My point is not about the script needing lots of flexibility. But rather if a particular project needs to adjust a script to fit the constraints of their project. e.g. if projectA needs the script changed for whatever reason, that shouldn't require it being changed for the shared profile. projectA should be able to overwrite the script if they need to.

The default script is still encapsulated in the shared profile, and only shows up in hiera of projectA if the project chooses to address it.

To me it seems much more flexible to not use static files if at all possible. I've moved away from using Puppet module files, and only use them as a last resort, occasionally  for binary files. Even most binary files I've switched to base64 encoding the binary content and supplying it via hiera.

Reply:
It's hard to predict the "IF's" that might come up.

Perhaps a best of both worlds approach is:
1. Move the file to "MODULE"/files or "MODULE"/templates (get's it out of hiera, provides an easy to manage file that is stored in a place that other maintainers will expect)
1. Add a class parameter(s) for "manage this file" (this would allow disabling puppet control of the file)
1. Add a class parameter(s) for "file content" (this would allow users of the module to override the contents if needed)

For items 2 & 3 above, they could be implemented as a single parameter of type Hash, with a key of 'filename' and sub keys that are 'manage' = [Boolean] and 'content' = [String]. If manage is True (the default) and 'content' is provided, then the user provided content will override the module provided file.

More so, item 1 could be the basic way to implement files inside profiles like this and items 2 & 3 could be a follow-on ticket to be created and implemented if needed, driven by a real use case from a project. That way we don't spend time trying to account for all possible "IF's" up front and save time by implementing overrides only in the cases where needed.